### PR TITLE
simple router with handler clojures support

### PR DESCRIPTION
**Pull request by StringNick**

### CODE REVIEW

**Comment by renerocksai** [comment]
Thanks, that's a great PR; I like the example. I hope I'll have more time later to look at it more closely before merging. I'd recommend naming it differently, more descriptively than hello3. Maybe something that communicates mapping routes to handler functions?

Also, please add it to the list of examples in the README, with a short description.

It's a really great example! Also, the `self` hack is genius! :+1:

File: src/router.zig [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+        if (route != null) {
+            return RouterError.AlreadyExists;
+        }
+
+        try self.routes.put(path, h);
+    }
+
+    pub fn serve(self: *Self, r: zap.Request) void {
+        var route = self.routes.get(r.path.?);
```
Comment (renerocksai) [inline src/router.zig:64]: I think r.path can be null, e.g. when requesting `http://server` as opposed to `http://server/`.

File: src/router.zig [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            return &run;
+        }
+
+        fn run(req: zap.Request) void {
+            hidden_func(hidden_self, req);
+        }
+    }).init(self, func);
+}
+
+pub const RouterOptions = struct {
```
Comment (renerocksai) [inline src/router.zig:27]: rename this to Options -> zap.Router.Options

File: src/router.zig [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    not_found: ?zap.HttpRequestFn = null,
+};
+
+pub const Router = struct {
```
Comment (renerocksai) [inline src/router.zig:31]: make this the main struct by removing line 31

File: src/zap.zig [hunk 4]
```diff
@@ -52,6 +52,8 @@ pub const Tls = @import("tls.zig");
 /// ```
 pub const Endpoint = @import("endpoint.zig");
 
+pub usingnamespace @import("router.zig");
```
Comment (renerocksai) [inline src/zap.zig:55]: pub const Router = ...

File: src/router.zig [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+const zap = @import("zap.zig");
+
+const Allocator = std.mem.Allocator;
+const RouterError = error{
+    AlreadyExists,
+    EmptyPath,
+};
+
+// inline closure for RequestFn with self argument
+pub inline fn RequestHandler(self: anytype, func: *const fn (@TypeOf(self), zap.Request) void) *const fn (zap.Request) void {
```
Comment (renerocksai) [inline src/router.zig:11]: maybe, if it should be available under zap namespace, move this into its own file and import in zap usingnamespace.

File: examples/hello3/hello3.zig [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+    try simpleRouter.handle_func("/geta", zap.RequestHandler(&somePackage, SomePackage.getA));
+
+    try simpleRouter.handle_func("/getb", zap.RequestHandler(&somePackage, SomePackage.getB));
+
+    try simpleRouter.handle_func("/inca", zap.RequestHandler(&somePackage, SomePackage.incrementA));
+
+    var listener = zap.HttpListener.init(.{
+        .port = 3000,
+        .on_request = zap.RequestHandler(&simpleRouter, &zap.Router.serve),
```
Comment (renerocksai) [inline examples/hello3/hello3.zig:95]: zap.Router.RequestHandler or zap.Requesthandler, depending on whether RequestHandler might be useful outside of zap.Router

**Review by renerocksai — CHANGES_REQUESTED** [review]
Please also document RequestHandler and Router.

**Comment by StringNick** [comment]
thank you for review and comments, i will fix it, first version name was SimpleRouter, but i saw u refactored names and renamed too, will use SimpleRouter

**Review by renerocksai — APPROVED** [review]
This is looking amazingly good!

**Comment by renerocksai** [comment]
Thx for the great work. Now, if you're up for it, could you please add doc comments for the zap.Router? Best take the zig-0.12.0 branch for that, as I fixed some const issues there (new zig is const-mad).

**Pull request merged**